README
This is Perl module Mojolicious::Plugin::Web::Auth.

INSTALLATION

Mojolicious::Plugin::Web::Auth installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Mojolicious::Plugin::Web::Auth

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Mojolicious::Plugin::Web::Auth documentation is available as in POD. So you can do:

    % perldoc Mojolicious::Plugin::Web::Auth

to read the documentation online with your favorite pager.

hayato imai


Changes
Revision history for Mojolicious-Plugin-Web-Auth

{{$NEXT}}

0.17 2019-04-21T15:08:44Z
        - merged PR #35. thanks Grinnz@github

0.16 2019-04-15T04:09:33Z
        - modify the required version to Mojolicious >= 7.13
        - merged PR #33 thanks oalders@github
        - merged PR #32. thanks oalders@github

0.15 2017-07-30T02:19:21Z
        - merged PR #30. thanks pablrod@github

0.14 2017-05-01T02:56:21Z
        - fixed issue #29. thanks guest20@github

0.13 2017-01-19T00:50:55Z
        - update pod. thanks oalders@github.
        - support refresh_token and access_type(Google). thanks ccakes@github.

0.12 2015-03-02T04:03:00Z
        - Minillaize
        - supported Mojo::Loader::load_class (added Mojolicious 5.81)

0.11 2014-10-09T01:24:56Z
        - added an authorize_header property in Mojolicious::Plugin::Web::Auth::OAuth2 to support a Authorization Header. thanks levonet@github.

0.10 2014-09-05T13:48:41Z
        - fixed to change the scheme, when the request-header has x-forwarded-proto value of "https". thanks shohey1226@github

0.09  2014-08-01 23:34:27 JST
        - fixed the error handling when user has canceled authentication. thanks sachinjsk@github
        - added the validate_state option in Mojolicious::Plugin::Web::Auth::OAuth2. thanks mala@github

0.08  2014-03-24 13:24:58 JST
        - fixed to support a content-type with charset. thanks uzulla@github.

0.07  2014-01-13 16:17:39 JST
        - fixed deprecated method (Mojo::UserAgent#name). thanks fayland@github, battlemidget@github.

0.06  2013-11-10 12:01:22 JST
        - fixed a bug where a module that is cached incompletely

0.05  2013-11-06 10:23:09 JST
        - fixed support for prefork server (Hypnotoad, Starman, Starlet, etc.)

0.04  2013-05-30 09:03:44 JST
        - added the dependancy of IO::Socket::SSL

0.03  2013-05-01 18:21:38 JST
        - removed the dependancy of Net::Twitter::Lite
        - removed the dependancy of WebService::Dropbox
        - millaize
        - updated the document

0.02
        - added  Mojolicious::Plugin::Web::Auth::Site::Instagram
        - chanded to use Mojo::UserAgent::post with Mojolicious 3.85 or higher

0.01
        - original version


lib/Mojolicious/Plugin/Web/Auth.pm
package Mojolicious::Plugin::Web::Auth;

use strict;
use warnings;

our $VERSION = '0.17';

use Mojo::Base 'Mojolicious::Plugin';

sub register {
    my ( $self, $app, $args ) = @_;
    my $module = delete $args->{module} or die "Missing mandatory parameter: module";
    my $klass  = join '::', __PACKAGE__, 'Site', $module;
    if ($Mojolicious::VERSION >= 5.81) {
        Mojo::Loader::load_class($klass);
    } else {
        Mojo::Loader->load($klass);
    }

    my $moniker           = $klass->moniker();
    my $authenticate_path = delete $args->{authenticate_path} || "/auth/${moniker}/authenticate";
    my $callback_path     = delete $args->{callback_path}     || "/auth/${moniker}/callback";

    # handlers
    my $on_finished = delete $args->{on_finished} or die "Missing mandatory parameter: on_finished";
    my $on_error    = delete $args->{on_error} || sub {
        my ( $c, $err ) = @_;
        die "Authentication error in $module: $err";
    };

    # auth object
    my $auth = $klass->new(%$args);

    $app->hook( before_dispatch => sub {
        my $c    = shift;
        my $path = $c->req->url->path;
        if ( $path->contains($authenticate_path) ) {
            my $forwarded_proto = $c->req->headers->header('x-forwarded-proto');
            $c->req->url->base->scheme('https') if (defined $forwarded_proto && $forwarded_proto eq 'https');
            my $callback = $c->req->url->path($callback_path)->to_abs;
            return $c->redirect_to( $auth->auth_uri( $c, $callback ) );
        }
        elsif ( $path->contains($callback_path) ) {
            return $auth->callback( $c, +{
                on_finished => sub {
                    $on_finished->($c, @_);
                },
                on_error => sub {
                    $on_error->($c, @_);
                },
            } );
        }
    } );

    return $self;
}

1;

__END__

=head1 NAME

Mojolicious::Plugin::Web::Auth - Authentication plugin for Mojolicious

=head1 SYNOPSIS

  # Mojolicious
  $self->plugin('Web::Auth',
      module      => 'Twitter',
      key         => 'Twitter consumer key',
      secret      => 'Twitter consumer secret',
      on_finished => sub {
          my ( $c, $access_token, $access_secret ) = @_;
          ...
      },
  );

  # Mojolicious::Lite
  plugin 'Web::Auth',
      module      => 'Twitter',
      key         => 'Twitter consumer key',
      secret      => 'Twitter consumer secret',
      on_finished => sub {
          my ( $c, $access_token, $access_secret ) = @_;
          ...
      };


  ### default authentication endpoint: /auth/{moniker}/authenticate
  # e.g.)
  # /auth/twitter/authenticate
  # /auth/facebook/authenticate
  ### default callback endpoint: /auth/{moniker}/callback
  # e.g.)
  # /auth/twitter/callback
  # /auth/facebook/callback

  # If, for example, your Instagram app needs more permissions than the
  # defaults allow:

  # Mojolicious
  $self->plugin('Web::Auth',
      module      => 'Instagram',
      key         => 'Instagram consumer key',
      secret      => 'Instagram consumer secret',
      on_finished => sub {
          my ( $c, $access_token, $access_secret ) = @_;
          ...
      },
      authorize_url => 'https://api.instagram.com/oauth/authorize?scope=public_content',
  );

=head1 DESCRIPTION

L<Mojolicious::Plugin::Web::Auth> is an authentication plugin for L<Mojolicious>.

=head1 METHODS

L<Mojolicious::Plugin::Web::Auth> inherits all methods from L<Mojolicious::Plugin>.

=head1 OPTIONS

L<Mojolicious::Plugin::Web::Auth> supports the following options.

=head2 C<module>

This is a module name for authentication plugins.

Dropbox, Facebook, Github, Google, Instagram, Twitter.

=head2 C<key>

Consumer key.

=head2 C<secret>

Consumer secret.

=head2 C<scope>

Optional: OAuth 2.0 only.

  # Facebook
  plugin 'Web::Auth',
      module      => 'Facebook',
      key         => 'Facebook App ID',
      secret      => 'Facebook App Secret',
      scope       => 'email,user_birthday',
      on_finished => sub {
          my ( $c, $access_token, $user_info ) = @_;
          ...
      };

=head2 C<validate_state>

Optional: OAuth 2.0 only. Default value is 1, see L<http://tools.ietf.org/html/rfc6819#section-5.3.5>.


=head2 C<on_finished>

  # Mojolicious::Lite
  plugin 'Web::Auth',
      module      => 'Twitter',
      key         => 'Twitter consumer key',
      secret      => 'Twitter consumer secret',
      on_finished => sub {
          my ( $c, $access_token, $access_secret, $user_ino ) = @_;
          ...
      };

This callback is run when authentication was finished.

=head3 arguments

=over 4

=item * OAuth 1.0(A)

Dropbox, Twitter, etc.

=over 4

=item Mojolicious::Controller

=item access_token

=item access_secret

=item user_info ( enabled 'user_info' )

=back

=item * OAuth 2.0

Facebook, Github, Google, Instagram, etc.

=over 4

=item Mojolicious::Controller

=item access_token

=item user_info ( enabled 'user_info' )

=item rest of the response data (HASHREF)

id_token, expires, refresh_token, etc.

=back

=back

=head2 C<on_error>

This is a callback is run on authentication errors.

=head1 AUTHOR

hayajo E<lt>hayajo@cpan.orgE<gt>

=head1 CONTRIBUTORS

Many thanks to the contributors for their work.

=over 4

=item FAYLAND

=item uzulla@github

=item sachinjsk@github

=item mala@github

=item shohey1226@github

=item levonet@github

=item oalders@github

=item ccakes@github

=back

=head1 COPYRIGHT

Copyright 2013- hayajo

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Mojolicious>, L<Amon2::Auth>

=cut
